### DISCUSSION TITLE
Typst Requires dylib loading?

### DISCUSSION THREAD

**Discussion by jtlandis** [quarto-dev/quarto-cli#9272] [Q&A]
### Description

I heard about `typst` recently and I'm simply trying to get a document to render with `typst` as the quarto format. I was under the impression that `typst` was completely bundled in `quarto` and it did not rely on external binaries. The following error I'm getting on a render attempt seems to imply otherwise.

````qmd
---
title: Typst
format: typst
---

Basic document to render
````
```
==> quarto preview typst-test.qmd --to typst --no-watch-inputs --no-browse

pandoc 
  to: typst
  output-file: typst-test.typ
  standalone: true
  shift-heading-level-by: -1
  default-image-extension: svg
  wrap: none
  citeproc: false
  
metadata
  title: Typst
  
[typst]: Compiling typst-test.typ to typst-test.pdf...dyld: Library not loaded: /usr/local/opt/xz/lib/liblzma.5.dylib
  Referenced from: /Applications/quarto/bin/tools/x86_64/typst
  Reason: no suitable image found.  Did find:
	/usr/local/opt/xz/lib/liblzma.5.dylib: code signature in (/usr/local/opt/xz/lib/liblzma.5.dylib) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed.
```

Has anyone seen an error like this before?

```
$ quarto --version
1.4.549
Rstudio: Version 2023.12.1+402 (2023.12.1+402)
```
Os Version: MacOS Big Sur (11.7.4)

**Comment by mcanouil** [comment]
How did you install Quarto?

I believe the issue is the installation here.
Maybe uninstall, cleanup remaining files, reinstall.

**Reply by jtlandis** [reply]
To be clear, do you mean to reinstall quarto by itself? or should I reinstall RStudio since quarto is packaged with it?

I have already reinstalled Rstudio, but I can try and download quarto instead.

**Reply by jtlandis** [reply]
Same issue when updating to quarto version `1.4.553` 😢

**Reply by mcanouil** [reply]
I don't think RStudio installs Quarto and less likely as an application `/Applications/quarto`.
RStudio does come with an embedded version, but you seem to have installed Quarto yourself, isn't it?

**Reply by jtlandis** [reply]
I have installed quarto myself. quarto for me is installed in `/usr/local/bin/quarto`

**Reply by mcanouil** [reply]
From the log you shared, it's in "Application".
I do suggest you remove all Quarto installs you made.
Ensure it remains only the one from RStudio which I believe would be located inside RStudio PATH.

**Reply by cscheid** [reply]
`/usr/local/opt/xz/lib/liblzma.5.dylib`

This is weird - I don't know why dyld would be trying to load that. We don't install anything in `/usr/local/opt`. Do you have third-party software that takes over `/usr/local/opt/`? Inspecting your environment might help.

It's also a _bit_ suspicious that this is about `liblzma`, in case you're following the recent [backdoor mess](https://www.openwall.com/lists/oss-security/2024/03/29/4). I wonder if some third party software you installed has put in mitigations for liblzma and it's interfering with typst somehow.

This is all pretty weird.

**Reply by jtlandis** [reply]
Sorry to be more clear, it is installed in `/Applications`

```
$ ls -l /usr/local/bin/quarto 
lrwxr-xr-x  1 root  admin  31 Apr  4 11:13 /usr/local/bin/quarto -> /Applications/quarto/bin/quarto
```
@cscheid I think I'm good to avoid all the backdoor mess

```
$ xz --version
xz (XZ Utils) 5.2.5
liblzma 5.2.5
```

I uninstalled the one in `/Applications`, now my quarto version through RStudio is `1.3.450`

**Reply by cscheid** [reply]
@jtlandis the point I was trying to make is that our bundled version of `typst` shouldn't be getting anywhere near `/usr/local/opt`. 

I can only assume there's something about your shell environment that's causing that to happen, and the fact that `libxzma` is involved is a bit suspicious to me. Not that you're vulnerable somehow, but that some third-party software in your system, that might be managing `/usr/local/opt` is doing something additional to your shell environment.

**Reply by cscheid** [reply]
Adding to this, the `typst` version that we ship in 1.4 doesn't seem to require `libzxma` at all, at least on aarch64 macOS:

```
$ cd QUARTO_1.4.459_INSTALL/bin/tools/aarch64
$ otool -L typst
typst:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1953.255.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1228.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

Arch Linux reports the required .so files for every package in their distro, and [their information for typst 0.11.0 is consistent with what I'm reporting](https://archlinux.org/packages/extra/x86_64/typst/sonames/).

**Reply by jtlandis** [reply]
I'll see if I can figure out whats going on.

I've done the following

```
<user>$ su <user> #new shell
bash-3.2$ PATH=/usr/bin:/usr/local/bin:/usr/sbin:/bin
bash-3.2$ quarto preview typst-test.qmd 
pandoc 
  to: typst
  output-file: typst-test.typ
  standalone: true
  shift-heading-level-by: -1
  default-image-extension: svg
  wrap: none
  citeproc: false
  
metadata
  title: Typst
  
[typst]: Compiling typst-test.typ to typst-test.pdf...dyld: Library not loaded: /usr/local/opt/xz/lib/liblzma.5.dylib
  Referenced from: /Applications/quarto/bin/tools/x86_64/typst
  Reason: image not found
 ```
 This is after downloading `quarto 1.4.553` for MacOS from the [quarto download site](https://quarto.org/docs/download/), it was installed to `/Applications/quarto`.

 Note the reason the output is slightly different is because I had removed `xz` from `/usr/local/opt/xz`

**Reply by mcanouil** [reply]
Does your Mac has an Intel or an Apple Silicon CPU?

**Reply by cscheid** [reply]
One thing you can try is the following. Add `keep-typ: true` to your document's metadata. Then, you'll have a `typst-test.typ` file in your file system after `quarto render`. From there, you can do

```
$ quarto typst compile typst-test.typ
```

which calls typst directly without any Quarto involvement. If that fails, then it's very likely something with your setup or with typst itself, and not something we can fix. You could also try to install typst yourself, and then attempt to compile `typst-test.typ` with the installed typst binary.

**Reply by jtlandis** [reply]
> Does your Mac has an Intel or an Apple Silicon CPU?

Its an Intel Mac (2017)

@cscheid I attempted your suggestion, and unfortunately, it did not work. I did have RUST installed locally so I installed typst through [their github](https://github.com/typst/typst/tree/main) using the `cargo install --git ...` command in their README. This installed typst to `~/.cargo/bin/typst`.
Calling 
```
$ quarto typst compile typst-test.typ
```
still failed with the same error. However calling
```
$ typst compile typst-test.typ
```
compiled the pdf. If I had second mac to test this on, I could give you all some more feedback, sorry this is the best I can do at the moment.

**Reply by cscheid** [reply]
As a temporary workaround, you should be able to add

`export QUARTO_TYPST=~/.cargo/bin/typst`

to your shell, and quarto will use your installed version of typst. I'm away from an Intel macOS machine right now but I'll try some things and follow up when I have access to one.
